File: day25/day25.cpp
#include <algorithm>
#include <fstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

struct Map
{
	array<string, 3> data;
	size_t width, height;
	unsigned cur;

	void single_step(int herd, unsigned dx, unsigned dy)
	{
		const string& src = data[cur];
		if (++cur == 3)
		{
			cur = 0;
		}
		string& dst = data[cur];

		fill(dst.begin(), dst.end(), '.');
		for (size_t y = 0, off = 0; y < height; y++)
		{
			for (size_t x = 0; x < width; x++, off++)
			{
				if (src.at(off) == '.')
				{
				}
				else if (src.at(off) != herd)
				{
					dst[off]= src.at(off);
				}
				else
				{
					size_t nx = x;
					if ((nx += dx) == width)
					{
						nx = 0;
					}
					size_t ny = y;
					if ((ny += dy) == height)
					{
						ny = 0;
					}
					size_t noff = ny * width + nx;
					if (src.at(noff) == '.')
					{
						dst[noff] = herd;
					}
					else
					{
						dst[off] = herd;
					}
				}
			}
		}
	}

	void step()
	{
		single_step('>', 1, 0);
		single_step('v', 0, 1);
	}

	bool deadlock() const
	{
		const string& a = data[cur];
		const string& b = data[(cur + 1) % 3];
		return !memcmp(a.data(), b.data(), a.size());
	}
};

istream& operator>>(istream& in, Map &m)
{
	m.width = m.height = 0;
	m.cur = 0;

	string line;
	while (getline(in, line))
	{
		if (!m.width)
		{
			m.width = line.size();
		}
		m.height++;
		for (auto c: line)
		{
			m.data[0].push_back(c);
		}
	}
	m.data[1].resize(m.width * m.height, '.');
	m.data[2].resize(m.width * m.height, '.');

	return in;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	Map m;
	in >> m;
	in.close();

	int i = 0;
	do
	{
		m.step();
		i++;
	} while (!m.deadlock());

	fmt::print("Part1: {}\n", i);
	return 0;
}

File: day25/day25.py
#!/usr/bin/env python3

with open("input", "rt") as f:
    txt = f.read()

class Map(object):
    def __init__(self, txt):
        self.m = [list(row) for row in txt.splitlines()]
        self.swap = None
        self.width, self.height = len(self.m[0]), len(self.m)

    def single_step(self, element, dx, dy):
        self.swap = [["." for x in range(self.width)] for y in range(self.height)]
        for y, row in enumerate(self.m):
            for x, e in enumerate(row):
                if e == ".":
                    pass
                elif e != element:
                    self.swap[y][x] = e
                else:
                    nx = (x + dx) % self.width
                    ny = (y + dy) % self.height
                    if self.m[ny][nx] == ".":
                        self.swap[ny][nx] = element
                    else:
                        self.swap[y][x] = element

        self.m = self.swap

    def step(self):
        old = self.m
        self.single_step(">", 1, 0)
        self.single_step("v", 0, 1)

        # deadlock check
        for y, row in enumerate(old):
            for x, e in enumerate(row):
                if e != self.m[y][x]:
                    return True
        return False

    def __str__(self):
        return "\n".join("".join(row) for row in self.m)

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        m = Map(txt)
        i = 0
        while True:
            i += 1
            if not m.step():
                break

        print("Part1:", i)
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)
